### ISSUE TITLE
OOPS! Hercules has crashed!

### ISSUE THREAD

**Issue by mcisho** [SDL-Hercules-390/hyperion#303]
Windows 10 virtual machine on a Linux system. Hercules source had been updated to the latest version and rebuilt earlier today. Hercules had been running a Linux s390x guest which had been shut down cleanly. Entered the exit command and Hercules crashed. These are the last few messages:-

```
17:24:52.033 00001584 HHC00101I Thread id 000021c8, prio 4, name 'cckd_gcol' ended
17:24:52.037 00001584 HHC00101I Thread id 00000468, prio 1, name 'cckd_writer thread 1' ended
17:24:52.037 00001584 HHC00101I Thread id 00001bb0, prio 1, name 'cckd_writer thread 2' ended
17:24:52.106 00001584 
17:24:52.106 00001584 
17:24:52.107 00001584                       ***************
17:24:52.107 00001584                       *    OOPS!    *
17:24:52.107 00001584                       ***************
17:24:52.107 00001584 
17:24:52.107 00001584                     Hercules has crashed!
```

The crash had happened with previous versions of Hercules, but I hadn't reported it yet. I have the dump and log, where would you like me to FTP them to?

Ian

**Comment by Fish-Git** [comment]
> I have the dump and log, where would you like me to FTP them to?

Absolutely! Config file and source code too if  you made any local changes. Thanks!

**Comment by Fish-Git** [comment]
QUESTION: what version of Visual Studio did you use to build Hercules with? VS2015? VS2017? Or VS2019? I'm guessing it was 2019, but I want to be sure. Thanks.

**Comment by mcisho** [comment]
Built with VS2017. There were no local code changes.

Can you remind me of the URL to FTP to, please? I thought I had it noted somewhere, I just can't find where.

Ian

**Comment by Fish-Git** [comment]
> Built with VS2017. There were no local code changes.

Ok. Thanks.

> Can you remind me of the URL to FTP to, please? I thought I had it noted somewhere, I just can't find where.

_(Oops!)_ I should have provided that. Sorry. Your questions regarding whether you should FTP the dump to me or not threw me off, causing me to presume you already knew the URL. I shouldn't have presumed like that. My bad. Here's the URL:

* ftp://www.softdevlabs.com
* ftp://www.softdevlabs.com/incoming/

Basically just connect to the FTP server at the default address (www.softdevlabs.com) and then look for a folder called `incoming`.

The `incoming` folder is a _**write-only**_ folder, so:

1. You can only _upload_ files into it, but you cannot download anything from it.
2. You won't be able to see any files already there (i.e. you won't be able to do a `dir` or `ls` of that folder to see its contents since obtaining a directory listing requires _read_ permission which the folder doesn't have).

_Please compress the dump and all accompanying files before uploading (e.g. config file, log file, etc)._ Most crash dumps contain a lot of binary zeros and compress quite well. My FTP space, while being quite large and able to hold many gigbytes, _**is**_ nonetheless limited. File type doesn't matter. .zip, .bz2, .gz, .tar, .7z, etc. is all fine. I can handle pretty much anything.

Please send me a private email (or post a comment here, it doesn't matter) whenever the upload is complete, and I'll take things from there. Thanks.

**Comment by mcisho** [comment]
Issue-303.zip uploaded.
Ian

**Comment by Fish-Git** [comment]
Ian (@mcisho),

I'm afraid I'm going to need you to upload something else for me too: your Hercules binaries and debugging symbols! -- i.e. the contents of your:

* C:\Users\IRS\Documents\HercSDL\Hyperion\msvc.AMD64.<b><i>bin</i></b>\
* C:\Users\IRS\Documents\HercSDL\Hyperion\msvc.AMD64.<b><i>pdb</i></b>\

directories.

Please .zip them and upload them to the same place as before: my FTP "incoming" folder.

I apologize for not mentioning that earlier. I forgot. Under normal circumstances I don't need people to do that, because, so far, all crash dumps that I've had to analyze have all been for crashes that have occurred with one of the official release versions or one of my pre-built snapshots, and I already have the exact binaries and exact debug symbols for those.

But you built Hercules for yourself on a Windows 10 system that I do not have, so I am unable to build the same exact set of binaries and debugging symbols that you have. Thus I need yours.

Once I get your Hercules binaries (.exe and .dll files) and debugging symbols (.pdb files), I will then be able to analyze your crash dump.

Thanks.

(And sorry about this!)

**Comment by mcisho** [comment]
Issue-303-2.zip uploaded.
Ian

**Comment by Fish-Git** [comment]
_Got it!_ Thanks. Analysis in progress...

**Comment by Fish-Git** [comment]
This is an interesting crash!

Initial analysis reveals the crash is occurring in fthreads where it's calling `TryEnterCriticalSection` as part of its processing of the `cckd_writer` thread's `release_lock( &cckdblk.wrlock )` request that it does immediately before the thread exits. The only problem is, the fthreads logic that does that is in the `hutil.dll` which _may_ _(unconfirmed!)_ have already been unloaded as a result of Hercules having already exited. But that's just a guess.

It appears a lot of our threading code -- especially our dasd code -- was not properly written to properly wait for all sub-threads that it may create to first exit before itself exits. As a result, it's possible _(in fact quite likely!)_ that some worker threads may still be running whenever Hercules shuts down and tries to exit.

Depending on how such threads are written and what they do, that's usually not a big problem. The operating system simply kills (terminates) such threads whenever the main process terminates.

But if any of those still-running-threads require the services of another DLL (`hutil.dll` in this particular case), that could likely be a problem. There is no guarantee as to the order in which DLLs are unloaded by the operating system when a process exits. That's why it's important for there to be a clearly defined shutdown (exit) sequence for any multi-threaded process. The main shutdown (process exit) logic should first wait for all of the threads that it created to exit first, before it can then safely exit. Same goes for any/all of the threads that THOSE threads might create. That is to say, each thread that creates a thread should wait for _**all**_ of the threads it created to first exit before it can then safely exit, etc.

That's the only way to be 100% certain that when the main process (main thread) exits, there are _ZERO_ stray threads running. When a process exits, there should be _**no other threads running!**_

But I don't think that's true with Hercules. I suspect there a lots of places throughout Hercules where threads are creating other threads but then exiting themselves without first waiting for the threads they created to exit first, leading to crashes such as this one.

But that's all just an educated guess based on my years of experience programming on Windows.

Now comes the question that I have to ask that everyone hates me asking _(including myself!):_

_**Is this problem (crash) reproducible?**_

Everything I've written is pure speculation (i.e. guesswork). There is not enough information in the dump to determine, postmortem, what the exact sequence of events were leading up to the crash. I can only see the end results. I cannot see what caused it. What sequence of events lead to the condition that caused the crash.

If the problem is easily reproducible (I myself have _never_ seen/had it occur on my own system in the years I've been involved with Hercules development, but then maybe I've just been incredibly luck) then I _might_ be able to look into the problem a bit deeper and _maybe_ positively confirm _(or deny!)_ my suspicions, but without a reproducible test case that's going to be difficult if not impossible.

In any case, someone should probably create a new GitHub "To Do" Issue for closely reviewing our entire startup and shutdown logic to ensure we're doing things right. _Especially_ the thread creation and exiting logic as I explained (which it where I suspect our problem lies).

That's it. There's not much more I can do. &nbsp;&lt;shrug&gt;

**Comment by Fish-Git** [comment]
For those interested, here's the full windbg `!analyze -v` output:

* [_!analyze -v.txt](https://github.com/SDL-Hercules-390/hyperion/files/4691128/_.analyze.-v.txt)

**Comment by mcisho** [comment]
It's certainly reproducible in my environment. I IPL'ed and shutdown z/VM (it's quicker than zLinux) 5 times, end each time I entered the exit command Hercules crashed. I hadn't used my Win10 virtual machine for some time, but I do remember the crash happening in the past. Unfortunately I don't remember exactly when the crashes started to happen. I always thought 'I don't have time just now, I'll report the crash the next time'.
Ian

**Comment by Fish-Git** [comment]
Interesting...

I noticed you have multiple networking devices defined _(one of each!)._ I'm now wondering if maybe that might have something to do with the crash (even though I'm not seeing anything for that in the dump) since I did make a commit recently to fix a bug in that area ().

What does your z/VM system config look like for all of those networking devices? (rdevice? vswitch?)

z/VM user direct too? (since I think that's also where you define such stuff too?)

I'm still a little weak regarding z/VM networking setup! I originally struggled a bit setting up my two z/VM test systems (one for LCS, the other for QETH), but that was a long time ago, and I have never tried modifying either of them since then. So I don't have one setup for CTCI. Is there anything I should know about that? Is there a trick to it? What's involved? Thanks.

_**OR...**_ maybe it would just be a lot faster and easier for you to send me _your_ z/VM dasds? There shouldn't be a lot of them and they shouldn't be that large, right? Would doing that maybe be possible? It would certainly speed things up a LOT if I had the exact same setup as you and could then easily reproduce your crash on demand!

I fear that trying to tweak my existing z/VM systems might take too much time and effort and might not even be successful in the end. It might just be much faster and easier for you to simply send me "your Hercules system" so to speak. Would that be possible?

**Comment by mcisho** [comment]
The CTCI isn't used, the first LCS is used by VM's TCP/IP stack, the other networking devices are used by virtual machines.

If I comment out all but the first LCS, Hercules always crashes on exit. If I comment out all of the networking devices, Hercules sometimes crashes on exit, maybe one time in three.

I will happily send you my VM disks, but I fear the speed or slowness of my environment is opening the timing hole which your environment won't.

**Comment by mcisho** [comment]
Issue-303-3.zip uploaded.
Ian

**Comment by Fish-Git** [comment]
> If I comment out all but the first LCS, Hercules always crashes on exit. If I comment out all of the networking devices, Hercules sometimes crashes on exit, maybe one time in three.

Hmmm. Okay. It doesn't sound like it has anything to do with the networking devices then. It's probably what I originally suspected: stray threads still running during exit.

> I will happily send you my VM disks, but I fear the speed or slowness of my environment is opening the timing hole which your environment won't.

You're probably right. I was going to tell you to forget it, but I see you've already uploaded it:

> Issue-303-3.zip uploaded.

Thanks. I'll give it a try. I suspect you're going to be right though: it's likely timing related so I probably won't be able to recreate it even using the exact same data (setup). Still, I'll give it a try. Thanks.

In the mean time, I'm _also_ going to try modifying the dasd code to properly wait for for all created threads to first exit before they exit too, and I'll post the patch for you to try when I have it ready.

I appreciate your patience with this, Ian. I know having a problem that isn't easily reproducible by others can be incredibly frustrating, so I appreciate your effort helping me to try and get to the bottom of it.

**Comment by Fish-Git** [comment]
Well, as expected, I could not reproduce your failure using your data, so I suspect the problem is very likely timing related, so let's try this:

Apply the following patch and rebuild and re-test:

```diff
--- hyperion-1/config.c	2020-05-18 12:26:39.051659700 -0700
+++ hyperion-0/config.c	2020-05-28 09:23:39.136229900 -0700
@@ -909,6 +909,8 @@
     broadcast_condition (&sysblk.ioqcond);
     release_lock (&sysblk.ioqlock);
 
+    SLEEP(1);
+
     /* release storage          */
     sysblk.lock_mainstor = 0;
     WRMSG( HHC01427, "I", "Main", !configure_storage(~0ULL) ? "" : "not " );
@@ -918,6 +920,9 @@
     WRMSG( HHC01427, "I", "Expanded", !configure_xstorage(~0ULL) ? "" : "not ");
 
     WRMSG(HHC01422, "I");
+
+    SLEEP(1);
+
 } /* end function release_config */
 
 #if defined( OPTION_SHARED_DEVICES )
--- hyperion-1/impl.c	2020-05-19 15:51:10.326333800 -0700
+++ hyperion-0/impl.c	2020-05-28 09:24:17.146283100 -0700
@@ -1340,7 +1340,7 @@
     socket_deinit();
 #endif
     fflush( stdout );
-    usleep( 10000 );
+    usleep( 1000000 );
     return 0; /* return back to bootstrap.c */
 } /* end function impl */
 
```

This is just a test patch and not the actual fix. I just want to see whether delaying the actual process exit prevents the crash from happening. If it does, then I can justify the extra effort involved in trying to actually fix the dasd code to properly wait for all of its threads to exit first.

If the above patch _doesn't_ workaround the problem, then my original suspicion was wrong and I'll have to scratch my head some more.

**Comment by mcisho** [comment]
Sadly the patch mad no difference. I've only tried one time, with an LCS defined, but exit resulted in an abend. These were the final log messages (no oops dialog-box appeared):-

20:56:42.609 00000F74 HHC00101I Thread id 000010b4, prio 4, name 'cckd_ra thread 1' ended
20:56:42.609 00000F74 HHC00101I Thread id 000010a4, prio 4, name 'cckd_ra thread 2' ended
20:56:42.621 00000F74 HHC00101I Thread id 00000ee0, prio 4, name 'cckd_gcol' ended
20:56:42.624 00000F74 HHC00101I Thread id 00001c0c, prio 1, name 'cckd_writer thread 1' ended
20:56:42.626 00000F74 HHC00101I Thread id 0000045c, prio 1, name 'cckd_writer thread 2' ended
20:56:42.638 00000F74 
20:56:42.638 00000F74 
20:56:42.638 00000F74                       ***************
20:56:42.638 00000F74                       *    OOPS!    *
20:56:42.639 00000F74                       ***************
20:56:42.640 00000F74 
20:56:42.640 00000F74                     Hercules has crashed!
20:56:42.640 00000F74 
20:56:42.641 00000F74 (you may or may not need to press ENTER if no 'oops!' dialog-box appears)
20:56:42.641 00000F74 
20:56:44.969 00000F74 HHC01427I Main storage released
20:56:44.969 00000F74 HHC01427I Expanded storage released
20:56:44.969 00000F74 HHC01422I Configuration released
20:56:46.073 00000F74 HHC01424I All termination routines complete
20:56:46.073 00000F74 HHC01425I Hercules shutdown complete
20:56:46.074 00000F74 HHC01412I Hercules terminated
20:56:46.074 00000F74 HHC00101I Thread id 000019bc, prio 7, name 'timer_thread' ended
20:56:46.133 00000F74 HHC01542I HDL: dyngui.dll terminated

Ian

**Comment by Fish-Git** [comment]
> Sadly the patch mad no difference.

Damn! &nbsp; `:(`

> (no oops dialog-box appeared)

_That's_ weird.

Just so we're on the same page, _how_ are you shutting down? Are you entering the z/VM `shutdown` command first, and then once the disabled wait message appears (in response to z/VM having quiesced the system), then entering the Hercules `exit` command? Or are you just entering the Hercules `exit` command and letting z/VM automatically detect that it's being shutdown?

In other words, are you able to successfully reach the "disabled wait" state with Hercules still up and running, and it only crashes when you _then_ enter the `exit` command? I'm curious as to whether the behavior is any different.

_(Dang! I HATE weird bugs like this!)_

**Comment by Fish-Git** [comment]
p.s. Remove the patch, rebuild and retest, and then send me your crash dump. Since you're now using z/VM (and not z/Linux like before), I want to see whether it's any different from your previous crash dump, or if it's still identifying the same problem in the same place as before (cckd dasd `cckd_writer` thread exit). Thanks!

**Comment by mcisho** [comment]
I'm not ignoring you, I'm just having trouble producing a dump! The Oops occurs every time I exit Hercules, the dialog box flashes before my eyes, but there is no dump being produced! Or at least I can't find it. I must have inadvertently changed something, I just don't know what.

**Comment by Fish-Git** [comment]
> The Oops occurs every time I exit Hercules, the dialog box flashes before my eyes, but there is no dump being produced!

_(SIGH)_ &nbsp; Fricking Windows... &nbsp; `:(`

**Comment by Fish-Git** [comment]
Ian (@mcisho),

_**BREAKTHROUGH!**_

It took all freaking day to do it, but I finally managed to recreate your crash using your dasds!

And I now (mostly) know what's causing it!

I unfortunately don't _(YET!)_ know _how_ it's occurring (i.e. the sequence of events that leads up to the bug), but I at least now know both _what_ and _why!_ (i.e. I've found the bug!) (mostly anyway)

The reason it took me all day was because I had to update my copy of Windows 10 first (running in a VMware virtual machine), which, thanks to Microsoft, is no longer a simply straightforward procedure (that and the fact that my Win10 system hadn't been updated in over a year, so it took a while, but after many reboots and much thumb twiddling waiting for Windows to do its thing I finally managed to complete it).

Then of course I had to uninstall VS2017 because, ONCE AGAIN, THANKS TO MICROSOFT, it no longer allowed me to use it! (apparently my "Microsoft account" that I used to "register" it with is "locked" and they (fricking Microsoft!) not only won't tell me WHY, but they won't let me unlock it either! (since they require a mobile phone they can send a text message to and I don't have a mobile phone!)) (I went through an alternate procedure but that didn't seem to work either, so I don't know what the frick I'm supposed to do now.)

I then installed VS2019 **`(*)`**, ... and git, ... and bunch of other tools, etc, that I use to do Hercules development, and then FINALLY, many, MANY hours later, finally managed to reach the point of being able to try your simple test.

And amazingly, it crashed! Just like you said!

AND, just like you later said, the "Oops!" message was displayed, but no dialog box appeared and no crash dump was created either! **`(**)`**

Anyway, it's late, I'm tired, and I need to get some sleep. I jump into trying to work out a fix tomorrow. I just wanted to let you know:

1. I found it! (more or less)

2. Forget trying to create a crash dump. It doesn't seem to work any longer Windows 10. (F**king M$!! I hated them a lot before but hate them even more after today!)

I'll post the details of the bug tomorrow (or maybe the next day, i.e. once I get to the root cause and  actually develop the fix for it).

That's all for now. More tomorrow (or the next day) after I get some much needed sleep!

Nighty-nite!

`(-.-)` &nbsp; Zzzzzz....
___
**`(*)`** Which I couldn't "register" for the same reason (locked account!), so I fear my VS2019 is also going to eventually stop working too in a few months, just like my VS2017 did. (sigh!)

**`(**)`**  It's probably something M$ f**ked up in their latest Win10 updates or something. They're becoming more and more incompetent with each passing month, I swear. It's depressing. AND annoying. AND frustrating. AND a thousand other things all at the same time... I hate it! &nbsp; `>8-<`

**Comment by Fish-Git** [comment]
_**CONFIRMED!**_

I have found and fixed the bug! &nbsp; `:-D`

_&lt;me: does happy dance&gt;_

Before I launch into describing the details of the bug however, please apply the following patch, rebuild and retest to confirm. Hercules should no longer crash for you.

_(Note: Windows CR/LF line endings)_

* [#303.patch.txt](https://github.com/SDL-Hercules-390/hyperion/files/4706739/303.patch.txt)

Please confirm for me that the above patch fixes the problem. Once confirmed, I'll go ahead and commit the fix and then describe the bug in detail. Thanks.

(Basically it was a race condition in cckd termination logic wherein a control block (`CCKDBLK`) was being cleared to zeros at shutdown before one or more of the various stlll-running cckd worker threads (read-ahead, writer and/or garbage-collector) had a chance to release their locks and completely exit.)

**Comment by mcisho** [comment]
Patch works just fine. Several exits, no Oops. Well done.

**Comment by Fish-Git** [comment]
Fixed by commit 5315756f79cfd86219362e9627215af87a140fe6.

As promised, for those interested, the bug (race condition) was in several places.

The first was in the `cckd_dasd_close_device` function where, near the very end of the function, it was doing:
```C
    /* If no more devices then perform global termination */
    if (cckdblk.dev1st == NULL) cckd_dasd_term ();
```
which is wrong, since the `cckdblk.dev1st` is always updated under the control of a lock. We can see this since just a few statements earlier it does:
```C
    /* Remove the device from the cckd queue */
    cckd_lock_devchain(1);
    if (dev == cckdblk.dev1st) cckdblk.dev1st = cckd->devnext;
    else
    {
        DEVBLK *dev2 = cckdblk.dev1st;
        CCKD_EXT *cckd2 = dev2->cckd_ext;
        while (cckd2->devnext != dev)
        {
           dev2 = cckd2->devnext; cckd2 = dev2->cckd_ext;
        }
        cckd2->devnext = cckd->devnext;
    }
    cckd_unlock_devchain();
```

The second, third and forth places were in the signaling logic of the three actual threads' termination. The `cckd_ra` asynchronous readahead thread, `cckd_writer` writer thread, and `cckd_gcol` garbage collection thread, which were all doing this:

```C
    /* Return with message if too many already started */
    if (writer > cckdblk.wrmax)
    {
        --cckdblk.wrs;  /* decrease threads started */
        --cckdblk.wra;  /* decrease threads active  */

        if (!cckdblk.wrmax)  /* choose thread termination message  */
        {
            signal_condition( &cckdblk.termcond );/* shutting down */

            if (!cckdblk.batch || cckdblk.batchml > 1)
              // "Thread id "TIDPAT", prio %d, name '%s' ended"
              LOG_THREAD_END( threadname  );
        }
        else
            if (!cckdblk.batch || cckdblk.batchml > 0)
                // "Ending thread "TIDPAT" %s, pri=%d, started=%d, max=%d exceeded"
                WRMSG( HHC00108, "W", TID_CAST( thread_id()), threadname,
                get_thread_priority(), writer, cckdblk.wrmax );

        release_lock( &cckdblk.wrlock );
        return NULL;
    }
```

and:

```C
    if (cckdblk.wrs == 0)
        signal_condition( &cckdblk.termcond );

    release_lock( &cckdblk.wrlock );
    return NULL;
} /* end thread cckd_writer */
```

As you can see, they were all doing the `signal_condition` while holding the lock, and then _after_ sending the signal, doing the `release_lock` afterwards.

As soon as the signal was sent, the `cckd_dasd_term` CCKD dasd global termination function, upon receiving the signal, then proceeded with global termination, eventually reaching the end of its function and doing:
```C
    memset(&cckdblk, 0, sizeof(CCKDBLK));
    return 0;
} /* end function cckd_dasd_term */
```
As you can see, it is clearing `CCKDBLK` to binary zeroes, which of course includes the `wrlock` and `ralock` and `gclock` locks which, depending on the timing, the thread(s) in question might still be holding and not released yet! Sometime later, _after_ `CCKDBLK` has already been cleared to zeroes, the host gets around to scheduling the  `cckd_writer` thread again (the one that sent the `cckdblk.termcond` signal), and as soon as it attempts to release it's `cckdblk.wrlock`, KABOOM!

The fix was to fix each thread to always release its lock _before_ issuing its `signal_condition` for `cckdblk.termcond`. (There is no requirement for the thread doing a `signal_condition` to hold the same lock that the thread(s) receiving the signal may have associated with the condition variable in the `wait_condition` call.) This ensures the locks are always released _before_ the `cckd_dasd_term` function clears `CCKDBLK` to binary zeroes, thereby preventing the crash.

That's it.

I should note that yes, it might have been much easier to simply remove the `memset(&cckdblk, 0, sizeof(CCKDBLK));` statement at the end of the  `cckd_dasd_term` function. That would certainly have prevented Hercules from crashing, but that to be the "**[band aid approach](https://www.google.com/search?q=define+band+aid+approach)**" IMHO, as it does not fix the actual underlying race condition. It only prevents the race condition from causing the crash. I prefer to treat the actual disease and not just the symptom (which is what the [band aid approach](https://www.google.com/search?q=define+band+aid+approach) does).

**Comment by Fish-Git** [comment]
Thank you, Ian, for all of your help resolving this issue (the dump, the dasds, etc). I could not have done it without you, buddy!
